aten/src/ATen/core/type.cpp

```diff
@@ -141,51 +141,39 @@ ListTypePtr ListType::ofBools() {
   return value;
 }
 
-c10::optional<TypePtr> tryEitherIsTheSuperType(const TypePtr& t1, const TypePtr& t2) {
-  if (t1->isSubtypeOf(t2)) {
-    return t2;
-  } else if (t2->isSubtypeOf(t1)) {
-    return t1;
-  } else {
-    return c10::nullopt;
-  }
-}
 
-c10::optional<TypePtr> unifyTypes(const TypePtr& t1, const TypePtr& t2) {
-  //cases that t1 == t2, or t1 is a type refinement of t2 and vice versa
-  if (auto maybe_supertype = tryEitherIsTheSuperType(t1, t2)) {
-    return *maybe_supertype;
-  }
-
-  // NB: we do not return NumberType because there is not currently enough
-  // operator support for it
+c10::optional<TypePtr> unifyTypes(const TypePtr& type1_input, const TypePtr& type2_input) {
+  auto t1 = type1_input;
+  auto t2 = type2_input;
 
+  // Handle non-container types which do not subtype each other and unify
   if (t1->kind() == TensorType::Kind && t2->kind() == TensorType::Kind) {
-    return t1->expect<TensorType>()->merge(t2->expect<TensorType>());
+    return type1_input->expect<TensorType>()->merge(type2_input->expect<TensorType>());
   }
 
-  if (t1->isSubtypeOf(TensorType::get()) && t2->isSubtypeOf(TensorType::get())) {
-    return static_cast<TypePtr>(TensorType::get());;
-  }
-
-  // if t1 is None and t2 is a concrete type, return Optional[t2] and vice versa
   if (t1->isSubtypeOf(NoneType::get()) && !t2->isSubtypeOf(NoneType::get())) {
     return OptionalType::create(t2);
   } else if (t2->isSubtypeOf(NoneType::get()) && !t1->isSubtypeOf(NoneType::get())) {
     return OptionalType::create(t1);
   }
 
-  //types which contain other types
-  if (t1->cast<ListType>() && t2->cast<ListType>()) {
-    // because we have runtime specializations of lists, e.g. int[] = std::vector<int64_t>
-    // int?[] = std::vector<IValue>  we don't allow type coercion,
-    // since t1 & t2 may have different runtime representations.
-
-    // allow Lists of different tensor types
-    auto unshaped_t1 = unshapedType(t1);
-    auto unshaped_t2 = unshapedType(t2);
-    return tryEitherIsTheSuperType(unshaped_t1, unshaped_t2);
-  } else if(t1->cast<TupleType>() && t2->cast<TupleType>()) {
+  // NB: we do not return NumberType because there is not currently enough
+  // operator support for it
+
+  // Handle bivariant type containers.
```

Comment:
Not sure if bivariant is correct in context, but I might be missing it. Regardless though, I think it'd probably be better just to note this logic implements the "See through optional" rule `unify(Optional[t1], t2) => Optional[unify(t1, t2)]` (resp `Optional[t2]`), in service of `T <: Optional[T]`.

Reply:
Yea i wasnt sure about this comment lol